backend/s3/s3.go

```diff
@@ -2078,15 +2110,15 @@ func (o *Object) uploadMultipart(ctx context.Context, req *s3.PutObjectInput, si
 	if concurrency < 1 {
 		concurrency = 1
 	}
-	bufs := make(chan []byte, concurrency)
+	concurrencyShares := make(chan struct{}, concurrency)
```

Comment:
You could use TokenDispenser here which is essentially this code packaged into a lib which will simplify the code a bit.

https://github.com/rclone/rclone/blob/481c8a40ea716bacb1f646604665f9859b65988f/lib/pacer/tokens.go#L10-L11